### ISSUE TITLE
C# ClientInterceptorContext Host always null

### ISSUE THREAD

**Issue by BenedicteLeo** [grpc/grpc#15884]
### What version of gRPC and what language are you using?
 C# 1.11.0
 
### What operating system (Linux, Windows, …) and version?
 Windows 10 x64
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
 C# .Net Framework 4.6.2
 
### What did you do?
I'm accessing the host in ClientInterceptorContext and it's always null:

```
public class ClientInterceptor : Interceptor
{
    public override AsyncUnaryCall<TResponse> AsyncUnaryCall<TRequest, TResponse>(
        TRequest request,
        ClientInterceptorContext<TRequest, TResponse> context,
        AsyncUnaryCallContinuation<TRequest, TResponse> continuation)
    {      
        string host = context.Host; // Null
        AsyncUnaryCall<TResponse> responseContinuation = continuation(request, context);

        return responseContinuation;
    }
}
```

 
### What did you expect to see?
 The target url that my channel is hitting.
 
### What did you see instead?
The host is always null.
 
This looks like a bug, unless I am misunderstanding the function of this field.

**Comment by mehrdada** [comment]
TLDR: `Host` is not the peer endpoint--it corresponds to the `Host` header in the HTTP request and to verify the endpoint's TLS certificate. It can be (and it more often than not is) `null`.

Sorry for the confusion: The documentation of the `Host` property probably needs to be clarified throughout the library (and not just in the interceptor API). I understand it might be a bit misleading.

`Host` is `null` by default and only gets populated if one calls the `WithHost` property on the generated client, which is a class derived from `ClientBase<T>` and invokes its `WithHost(string host)` method, resulting in a new client instance that will populate the Host property in the interceptor chain and eventually set the `CallInvocationDetails.Host` property which will eventually get passed down to gRPC 

Core as the `host` argument to `grpc_channel_create_call`.
(API: https://github.com/grpc/grpc/blob/e121fef19efdac51c20a4d0dab352b7ada387c72/include/grpc/grpc.h#L194 impl: 
https://github.com/grpc/grpc/blob/a789ac30673f6e56960329e802bc199d5a9ac760/src/core/lib/surface/channel.cc#L312)

The host argument is mostly used to change the SSL target name for SNI and certificate validation purposes. c.f. https://github.com/grpc/grpc/issues/2635#issuecomment-126957809

**Comment by BenedicteLeo** [comment]
Thank you for investigating this!
